meteo/_hycom.py

```diff
@@ -0,0 +1,175 @@
+
+from ._constants import HYCOM_MAX_DEPTH_INDEX
+from ._constants import POTENTIAL_TEMP_CORRECTION
+from ._constants import SUB_SAMPLE
+from ._literals import L_Days
+from ._literals import L_Months
+from ._utils import normalize_longitude
+
+logger = logging.getLogger(__name__)
+
+
+def ConvertTemp(salt, temp, dep):
+    # nz = temp.shape[0]
+    # ny = temp.shape[1]
+    # nx = temp.shape[2]
+    import gsw
+
+    pr = np.ones(temp.shape)
+    pre = pr * dep[:, None, None]
+    Pr = np.zeros(temp.shape)
+    ptemp = gsw.pt_from_t(salt, temp, pre, Pr) * POTENTIAL_TEMP_CORRECTION
+    return ptemp
+
+
+def get_database(date, Bbox=None):
+    if date > pd.Timestamp(2024, 8, 10):
+        year = date.year
+        str_ = (date-pd.Timedelta(days=1)).strftime("%Y%m%d")
+        # we pull the previous run after 12 hours (model start is always at midday), hence 'YYMMDD12_t0012' format
+        database = f"datasets/ESPC-D-V02/data/archive/{year}/US058GCOM-OPSnce.espc-d-031-hycom_fcst_glby008_{str_}12_t0012"
+    elif date >= pd.Timestamp(2018, 12, 4) and date <= pd.Timestamp(2024, 8, 10):
+        database = "GLBy0.08/expt_93.0"
+    elif date >= pd.Timestamp(2018, 1, 1) and date < pd.Timestamp(2018, 12, 4):
+        database = "GLBv0.08/expt_93.0"
+    elif date >= pd.Timestamp(2017, 10, 1) and date < pd.Timestamp(2018, 1, 1):
+        database = "GLBv0.08/expt_92.9"
+    elif date >= pd.Timestamp(2017, 6, 1) and date < pd.Timestamp(2017, 10, 1):
+        database = "GLBv0.08/expt_57.7"
+    elif date >= pd.Timestamp(2017, 2, 1) and date < pd.Timestamp(2017, 6, 1):
+        database = "GLBv0.08/expt_92.8"
+    elif date >= pd.Timestamp(2016, 5, 1) and date < pd.Timestamp(2017, 2, 1):
+        database = "GLBv0.08/expt_57.2"
+    elif date >= pd.Timestamp(2016, 1, 1) and date < pd.Timestamp(2016, 5, 1):
+        database = "GLBv0.08/expt_56.3"
+    elif date >= pd.Timestamp(1994, 1, 1) and date < pd.Timestamp(2016, 1, 1):
+        database = f"GLBv0.08/expt_53.X/data/{date.year}"
+    else:
+        raise ValueError(f"No data fro {date}!")
+    return database
+
+
+def get_idxs(date, database):
+    if date >= pd.Timestamp.now():
+        raise Exception("select time before today")
+    else:
+        if date >= pd.Timestamp(2024,8,10):
+            baseurl = f"https://tds.hycom.org/thredds/dodsC/{database}_s3z.nc?lat[0:1:-1],lon[0:1:-1],time[0:1:-1],depth[0:1:-1]"
+        else:
+            baseurl = f"https://tds.hycom.org/thredds/dodsC/{database}?lat[0:1:-1],lon[0:1:-1],time[0:1:-1],depth[0:1:-1]"
+
+    ds = xr.open_dataset(baseurl)
+    lon = ds.lon.data
+    lat = ds.lat.data
+    times = ds.time.data
+    ds.close()
+
+    time_idx = np.where(date == times)[0]
+    if len(time_idx) == 0:
+        raise Exception(f"No data for this date: {date}")
+
+    return time_idx[0], 0, len(lon)-1, 0, len(lat)-1
+
+def download_hycom(
+    year: int,
+    month: L_Months,
+    day: Annotated[L_Days, Parameter(show_choices=False)],
```

Comment:
OK, that's why cyclopts was imported. `day: L_Days` is the correct type hint, here.